### ISSUE TITLE
Excel Range Limitation when using ExcelSource?

### ISSUE THREAD

**Issue by xdororo** [roadrunnerlenny/etlbox#52]
Hi

I'm using Excel data source to upload an excel file, but I cannot read the values correctly when data column exceeds 20. For example as below, all excel columns after 20 return null. May I know if there's any limitation of Excel range regards to Excel source? Thanks.

Here is what I declare my excel range: 
Range = new ExcelRange(1, 2), 

My Class
{
...
[ExcelColumn(21)]
public string xxx { get; set; } 
}

**Comment by anlennar** [comment]
Hi @xdororo ,
thanks for your feedback and using ETLBox.
No, there is no limitation regarding the number of excel columns when using the Excel source. Your example code should work as expected. Unfortunately, you found a bug, which is the reason why the ExcelSource did not read all columns.
I was able to reproduce this bug within my tests. Luckily, this was easy to fix. The fix will be part of the release.
The next version is not fully ready yet, **but I already created a pre-release version which contains the bugfix:** 
https://www.nuget.org/packages/ETLBox/1.8.0-rc1

Please let me know if the pre-release and the bugfix works for you. 

Note: _This release does contain some minor breaking changes: I adjusted the names for CSVSource/CSVDestination and DBSource/DBDestination to real CamelCase - they are now CsvSource/CsvDestination and DbSource/DbDestination. ConnectionString was renamed to SqlConnectionString._

**Comment by xdororo** [comment]
Hi Anlennar

It's working fine with pre-release version of 1.8.0-rc1.
Really appreciate of your prompt help :)

**Comment by anlennar** [comment]
Good to hear! Let me know if you encounter any other issues.

**Comment by xdororo** [comment]
Hi there

Actually I am facing an issue handling blank rows from Excel which are
actually deleted. (the Excel source is reading a few rows as NULL values in
all the columns.)

Is there a way within ETLBOX can detect the end of row in the Excel sheet?
I appreciate any of the comment.

Brgds,
Melven



Andreas Lennartz <notifications@github.com> 於 2020年2月16日 週日 04:02 寫道：

**Comment by roadrunnerlenny** [comment]
Hi Melven,

I am not quite sure if I understand you correctly. The current ExcelSource is based on 
https://github.com/ExcelDataReader/ExcelDataReader
It is supposed to stop reading when it reads the end of the excel data. 
Could you give an example data file or a description how to reproduce the error?

**Comment by xdororo** [comment]
Hi Lenny

I will try to reproduce another  file to you because the file with errors
on my hand is kind of sensitive.

Excel source is doing well if I removed blank rows before loading data.

https://www.tech-recipes.com/rx/2189/excel_2007_eliminate_blank_rows/

roadrunnerlenny <notifications@github.com> 於 2020年2月17日 週一 23:57 寫道：

**Comment by roadrunnerlenny** [comment]
Hi Melven,
thanks for the feedback - glad to hear that deleting the blank rows manually could make it work. But of course it would be good if the excel source would behave as expected - looking forward to receive your excel demo file, so that I could reproduce this scenario in my test environment.

**Comment by xdororo** [comment]
Hi Anlennar

I attached two files which you could reproduce the issues. 
for the NG file, I got DBNull error when I load data from Excel Source. 
[Range = new ExcelRange(1, 3);]

If I manually delete the rows from row 9 to 1048576 (actually they are all blank cells), it works definitely. However in the most cases, we got excel files from another system, it's kind of annoying to ask users to do the check and manually delete the blank rows again before uploading the file.

I appreciate if this could be taken into consideration within ETLBOX because I really like it :).
Thanks.

[demo - excel blank rows - OK.xlsx](https://github.com/roadrunnerlenny/etlbox/files/4217554/demo.-.excel.blank.rows.-.OK.xlsx)
[demo - excel blank rows - NG.xlsx](https://github.com/roadrunnerlenny/etlbox/files/4217555/demo.-.excel.blank.rows.-.NG.xlsx)

Brgds,
Melven

**Comment by roadrunnerlenny** [comment]
Hi Melven,

thanks for the files. I was able to reproduce the blank lines and were able to implement a fix for this. They will now be suppressed, so they won't be forwarded to the connected components. I added a logic that basically filter out lines that are completely blank. 
Unfortunately, there are two downside here: The ExcelDataReader will still read try to read these lines, even though they are empty. So if there are *many* empty lines, it could take a little longer to read the full excel file. 
Also, blank lines will be ignore if they occur somewhere between your data rows. But I think this is kind of an expected behaviour, so I left it like this. 

 The changes will be part of the next release (1.8.0) which I will release in the next days.
